src/nvim/diff.c

```diff
@@ -745,37 +745,32 @@ static int diff_write_buffer(buf_T *buf, diffin_T *din)
     }
     return FAIL;
   }
-  din->din_mmfile.ptr = ptr;
-  din->din_mmfile.size = len;
+  m->ptr = ptr;
+  m->size = (long)len;
 
   len = 0;
   for (linenr_T lnum = 1; lnum <= buf->b_ml.ml_line_count; lnum++) {
-    for (char *s = ml_get_buf(buf, lnum, false); *s != NUL;) {
-      if (diff_flags & DIFF_ICASE) {
-        int c;
+    char *s = ml_get_buf(buf, lnum, false);
+    if (diff_flags & DIFF_ICASE) {
+      while (*s != NUL) {
         char cbuf[MB_MAXBYTES + 1];
 
-        if (*s == NL) {
-          c = NUL;
-        } else {
-          // xdiff doesn't support ignoring case, fold-case the text.
-          c = utf_ptr2char(s);
-          c = utf_fold(c);
-        }
+        // xdiff doesn't support ignoring case, fold-case the text.
+        int c = *s == NL ? NUL : utf_fold(utf_ptr2char(s));
         const int orig_len = utfc_ptr2len(s);
-        if (utf_char2bytes(c, cbuf) != orig_len) {
-          // TODO(Bram): handle byte length difference
-          memmove(ptr + len, s, (size_t)orig_len);
-        } else {
-          memmove(ptr + len, cbuf, (size_t)orig_len);
-        }
 
+        // TODO(Bram): handle byte length difference
+        char *s1 = (utf_char2bytes(c, cbuf) != orig_len) ? s : cbuf;
+        memmove(ptr + len, s1, (size_t)orig_len);
         s += orig_len;
-        len += orig_len;
-      } else {
-        ptr[len++] = *s == NL ? NUL : *s;
```

Comment:
Because copying one character at a time is such a good idea.

Reply:
I've benchmarked this, and this change is about 15% faster on MacOS with an M1 Pro.